File: test/javascript/testMultiSigWalletWithDailyLimit.js
const MultiSigWalletWithDailyLimit = artifacts.require('MultiSigWalletWithDailyLimit')
const web3 = MultiSigWalletWithDailyLimit.web3
const deployMultisig = (owners, confirmations, limit) => {
    return MultiSigWalletWithDailyLimit.new(owners, confirmations, limit)
}

const utils = require('./utils')
const ONE_DAY = 24*3600

contract('MultiSigWalletWithDailyLimit', (accounts) => {
    let multisigInstance
    const dailyLimit = 3000
    const requiredConfirmations = 2

    beforeEach(async () => {
        multisigInstance = await deployMultisig([accounts[0], accounts[1]], requiredConfirmations, dailyLimit)
        assert.ok(multisigInstance)
    })

    it('create multisig', async () => {
        const deposit = 10000
        
        // Send money to wallet contract
        await new Promise((resolve, reject) => web3.eth.sendTransaction({to: multisigInstance.address, value: deposit, from: accounts[0]}, e => (e ? reject(e) : resolve())))
        const balance = await utils.balanceOf(web3, multisigInstance.address)
        assert.equal(balance.valueOf(), deposit)
        assert.equal(dailyLimit, await multisigInstance.dailyLimit())
        assert.equal(dailyLimit, await multisigInstance.calcMaxWithdraw())

        // Withdraw daily limit
        const value1 = 2000
        let owner1Balance = await utils.balanceOf(web3, accounts[0])
        await multisigInstance.submitTransaction(accounts[0], value1, "", {from: accounts[1]})
        assert.equal(
            owner1Balance.add(value1).toString(),
            (await utils.balanceOf(web3, accounts[0])).toString()
        )
        assert.equal(
            balance.sub(value1).toString(),
            (await utils.balanceOf(web3, multisigInstance.address)).toString()
        )

        // Update daily limit
        const dailyLimitUpdated = 2000
        const dailyLimitEncoded = multisigInstance.contract.changeDailyLimit.getData(dailyLimitUpdated)
        const transactionId = utils.getParamFromTxEvent(
            await multisigInstance.submitTransaction(multisigInstance.address, 0, dailyLimitEncoded, {from: accounts[0]}),
            'transactionId', null, 'Submission')
        
        await multisigInstance.confirmTransaction(transactionId, {from: accounts[1]})
        assert.equal(dailyLimitUpdated, await multisigInstance.dailyLimit())
        assert.equal(0, await multisigInstance.calcMaxWithdraw())

        await utils.increaseTimestamp(web3, ONE_DAY+1)
        assert.equal(dailyLimitUpdated, (await multisigInstance.calcMaxWithdraw()).toNumber())

        // Withdraw daily limit
        const value2 = 1000
        owner1Balance = await utils.balanceOf(web3, accounts[0])
        await multisigInstance.submitTransaction(accounts[0], value2, "", {from: accounts[1]})
        assert.equal(
            owner1Balance.add(value2).toString(),
            (await utils.balanceOf(web3, accounts[0])).toString()
        )
        assert.equal(
            deposit-value2-value1,
            await utils.balanceOf(web3, multisigInstance.address)
        )
        assert.equal(
            dailyLimitUpdated - value2,
            await multisigInstance.calcMaxWithdraw()
        )
        await multisigInstance.submitTransaction(accounts[0], value2, "", {from: accounts[1]})
        assert.equal(
            owner1Balance.add(value2*2).toString(),
            (await utils.balanceOf(web3, accounts[0])).toString()
        )
        assert.equal(
            deposit-value2*2-value1,
            await utils.balanceOf(web3, multisigInstance.address)
        )
        assert.equal(
            dailyLimitUpdated - value2*2,
            await multisigInstance.calcMaxWithdraw()
        )

        // Third time fails, because daily limit was reached
        const transactionIdFailed = utils.getParamFromTxEvent(
            await multisigInstance.submitTransaction(accounts[0], value2, "", {from: accounts[1]}),
            'transactionId', null, 'Submission')
        assert.equal(false, (await multisigInstance.transactions(transactionIdFailed))[3])
        assert.equal(
            owner1Balance.add(value2*2).toString(),
            (await utils.balanceOf(web3, accounts[0])).toString()
        )
        assert.equal(
            deposit-value2*2-value1,
            await utils.balanceOf(web3, multisigInstance.address)
        )
        assert.equal(
            0,
            await multisigInstance.calcMaxWithdraw()
        )
        
        // Let one day pass
        await utils.increaseTimestamp(web3, ONE_DAY+1)
        assert.equal(dailyLimitUpdated, (await multisigInstance.calcMaxWithdraw()).toNumber())

        // Execute transaction should work now but fails, because it is triggered from a non owner address
        utils.assertThrowsAsynchronously(
            () => multisigInstance.executeTransaction(transactionIdFailed, {from: accounts[9]})
        )
        // Execute transaction also fails if the sender is a wallet owner but didn't confirm the transaction first
        utils.assertThrowsAsynchronously(
            () => multisigInstance.executeTransaction(transactionIdFailed, {from: accounts[0]})
        )
        // But it works with the right sender
        await multisigInstance.executeTransaction(transactionIdFailed, {from: accounts[1]})
        assert.ok((await multisigInstance.transactions(transactionIdFailed))[3])

        // Let one day pass
        await utils.increaseTimestamp(web3, ONE_DAY+1)
        assert.equal(
            dailyLimitUpdated,
            await multisigInstance.calcMaxWithdraw()
        )

        // User wants to withdraw more than the daily limit. Withdraw is unsuccessful.
        const value3 = 3000
        owner1Balance = await utils.balanceOf(web3, accounts[0])
        await multisigInstance.submitTransaction(accounts[0], value3, "", {from: accounts[1]})

        // Wallet and user balance remain the same.
        assert.equal(
            owner1Balance,
            (await utils.balanceOf(web3, accounts[0])).toString()
        )
        assert.equal(
            deposit-value2*3-value1,
            await utils.balanceOf(web3, multisigInstance.address)
        )
        assert.equal(
            dailyLimitUpdated,
            await multisigInstance.calcMaxWithdraw()
        )

        // Daily withdraw is possible again
        await multisigInstance.submitTransaction(accounts[0], value2, "", {from: accounts[1]})

        // Wallet balance decreases and user balance increases.
        assert.equal(
            owner1Balance.add(value2),
            (await utils.balanceOf(web3, accounts[0])).toString()
        )
        assert.equal(
            deposit-value2*4-value1,
            await utils.balanceOf(web3, multisigInstance.address)
        )
        assert.equal(
            dailyLimitUpdated-value2,
            await multisigInstance.calcMaxWithdraw()
        )
        // Try to execute a transaction tha does not exist fails
        const unknownTransactionId = 999
        utils.assertThrowsAsynchronously(
            () => multisigInstance.executeTransaction(unknownTransactionId, {from: accounts[0]})
        )
    })
})

File: test/python/test_multisig_wallet_with_daily_limit.py
# ethereum
from ethereum import tester as t
from ethereum.tester import keys, accounts, TransactionFailed, ContractCreationFailed
from preprocessor import PreProcessor
# standard libraries
from unittest import TestCase


class TestContract(TestCase):
    """
    run test with python -m unittest tests.test_multisig_wallet_with_daily_limit
    """

    HOMESTEAD_BLOCK = 1150000
    TWENTY_FOUR_HOURS = 86400  # 24h

    def __init__(self, *args, **kwargs):
        super(TestContract, self).__init__(*args, **kwargs)
        self.s = t.state()
        self.s.block.number = self.HOMESTEAD_BLOCK
        t.gas_limit = 4712388
        self.pp = PreProcessor()

    def test(self):
        # Create wallet
        required_accounts = 2
        daily_limit = 3000
        wa_1 = 1
        wa_2 = 2
        wa_3 = 3
        constructor_parameters = (
            [accounts[wa_1], accounts[wa_2], accounts[wa_3]],
            required_accounts,
            daily_limit
        )
        gas = self.s.block.gas_used
        self.multisig_wallet = self.s.abi_contract(
            self.pp.process('MultiSigWalletWithDailyLimit.sol', contract_dir='solidity/', add_dev_code=True),
            language='solidity',
            constructor_parameters=constructor_parameters,
            contract_name="MultiSigWalletWithDailyLimit"
        )
        self.assertLess(self.s.block.gas_used - gas, 2000000)
        print "Deployment costs: {}".format(self.s.block.gas_used - gas)
        # Create ABIs
        multisig_abi = self.multisig_wallet.translator
        # Send money to wallet contract
        deposit = 10000
        self.s.send(keys[wa_1], self.multisig_wallet.address, deposit)
        self.assertEqual(self.s.block.get_balance(self.multisig_wallet.address), deposit)
        self.assertEqual(self.multisig_wallet.dailyLimit(), daily_limit)
        self.assertEqual(self.multisig_wallet.calcMaxWithdraw(), daily_limit)
        # Withdraw daily limit
        value_1 = 2000
        wa_1_balance = self.s.block.get_balance(accounts[wa_1])
        self.multisig_wallet.submitTransaction(accounts[wa_1], value_1, "", sender=keys[wa_2])
        self.assertEqual(self.s.block.get_balance(self.multisig_wallet.address), deposit - value_1)
        self.assertEqual(self.s.block.get_balance(accounts[wa_1]), wa_1_balance + value_1)
        # Update daily limit
        daily_limit_updated = 2000
        update_daily_limit = multisig_abi.encode("changeDailyLimit", [daily_limit_updated])
        transaction_id = self.multisig_wallet.submitTransaction(self.multisig_wallet.address, 0,
                                                                update_daily_limit, sender=keys[wa_1])
        self.multisig_wallet.confirmTransaction(transaction_id, sender=keys[wa_2])
        self.assertEqual(self.multisig_wallet.dailyLimit(), daily_limit_updated)
        self.assertEqual(self.multisig_wallet.calcMaxWithdraw(), 0)
        self.s.block.timestamp += self.TWENTY_FOUR_HOURS + 1
        self.assertEqual(self.multisig_wallet.calcMaxWithdraw(), daily_limit_updated)
        # Withdraw daily limit
        value_2 = 1000
        wa_1_balance = self.s.block.get_balance(accounts[wa_1])
        self.multisig_wallet.submitTransaction(accounts[wa_1], value_2, "", sender=keys[wa_2])
        self.assertEqual(self.s.block.get_balance(self.multisig_wallet.address), deposit - value_1 - value_2)
        self.assertEqual(self.s.block.get_balance(accounts[wa_1]), wa_1_balance + value_2)
        self.assertEqual(self.multisig_wallet.calcMaxWithdraw(), daily_limit_updated - value_2)
        self.multisig_wallet.submitTransaction(accounts[wa_1], value_2, "", sender=keys[wa_2])
        self.assertEqual(self.s.block.get_balance(self.multisig_wallet.address), deposit - value_1 - value_2*2)
        self.assertEqual(self.s.block.get_balance(accounts[wa_1]), wa_1_balance + value_2*2)
        self.assertEqual(self.multisig_wallet.calcMaxWithdraw(), daily_limit_updated - value_2*2)
        # Third time fails, because daily limit was reached
        transaction_id = self.multisig_wallet.submitTransaction(accounts[wa_1], value_2, "", sender=keys[wa_2])
        self.assertFalse(self.multisig_wallet.transactions(transaction_id)[3])
        self.assertEqual(self.s.block.get_balance(self.multisig_wallet.address), deposit - value_1 - value_2*2)
        self.assertEqual(self.s.block.get_balance(accounts[wa_1]), wa_1_balance + value_2*2)
        self.assertEqual(self.multisig_wallet.calcMaxWithdraw(), 0)
        # Let one day pass
        self.s.block.timestamp += self.TWENTY_FOUR_HOURS + 1
        self.assertEqual(self.multisig_wallet.calcMaxWithdraw(), daily_limit_updated)
        # Execute transaction should work now but fails, because it is triggered from a non owner address
        self.assertRaises(TransactionFailed, self.multisig_wallet.executeTransaction, transaction_id, sender=keys[9])
        # Execute transaction also fails if the sender is a wallet owner but didn't confirm the transaction first
        self.assertRaises(TransactionFailed, self.multisig_wallet.executeTransaction, transaction_id, sender=keys[wa_1])
        # But it works with the right sender
        self.multisig_wallet.executeTransaction(transaction_id, sender=keys[wa_2])
        self.assertTrue(self.multisig_wallet.transactions(transaction_id)[3])
        # Let one day pass
        self.s.block.timestamp += self.TWENTY_FOUR_HOURS + 1
        self.assertEqual(self.multisig_wallet.calcMaxWithdraw(), daily_limit_updated)
        # User wants to withdraw more than the daily limit. Withdraw is unsuccessful.
        value_3 = 3000
        wa_1_balance = self.s.block.get_balance(accounts[wa_1])
        self.multisig_wallet.submitTransaction(accounts[wa_1], value_3, "", sender=keys[wa_2])
        # Wallet and user balance remain the same.
        self.assertEqual(self.s.block.get_balance(self.multisig_wallet.address), deposit - value_1 - value_2*3)
        self.assertEqual(self.s.block.get_balance(accounts[wa_1]), wa_1_balance)
        self.assertEqual(self.multisig_wallet.calcMaxWithdraw(), daily_limit_updated)
        # Daily withdraw is possible again
        self.multisig_wallet.submitTransaction(accounts[wa_1], value_2, "", sender=keys[wa_2])
        # Wallet balance decreases and user balance increases.
        self.assertEqual(self.s.block.get_balance(self.multisig_wallet.address), deposit - value_1 - value_2*4)
        self.assertEqual(self.s.block.get_balance(accounts[wa_1]), wa_1_balance + value_2)
        self.assertEqual(self.multisig_wallet.calcMaxWithdraw(), daily_limit_updated - value_2)
        # Try to execute a transaction tha does not exist fails
        transaction_id = 999
        self.assertRaises(TransactionFailed, self.multisig_wallet.executeTransaction, transaction_id, sender=keys[wa_1])
